libs/ui-watt/src/lib/components/clipboard/watt-copy-to-clipboard.directive.ts

```diff
@@ -0,0 +1,73 @@
+/**
+ * @license
+ * Copyright 2020 Energinet DataHub A/S
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License2");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  Directive,
+  ElementRef,
+  HostBinding,
+  HostListener,
+  Input,
+} from '@angular/core';
+import { Clipboard } from '@angular/cdk/clipboard';
+import { WattToastComponent, WattToastService } from '../toast';
+import da from './i18n/da.json';
+import en from './i18n/en.json';
+import { WattTranslationService } from '../../utils/translation';
+
+@Directive({
+  standalone: true,
+  selector: '[wattCopyToClipboard]',
+})
+export class WattCopyToClipboardDirective {
+  @Input('wattCopyToClipboard') text?: string;
+
+  constructor(
+    private element: ElementRef,
+    private clipboard: Clipboard,
+    private toast: WattToastService,
+    private translation: WattTranslationService
+  ) {
+    this.translation.addTranslation('clipboard', en, 'en');
+    this.translation.addTranslation('clipboard', da, 'da');
+  }
+
+  @HostBinding('style.cursor')
+  cursor = 'pointer';
+
+  @HostListener('click')
+  onClick() {
+    const success = this.text
+      ? this.clipboard.copy(this.text)
+      : this.clipboard.copy(this.element.nativeElement.innerText);
+
+    if (success) {
+      this.toast.open({
+        type: 'success',
+        message: this.translation.translate('clipboard.success'),
+      });
+    } else {
+      this.toast.open({
+        type: 'danger',
+        message: this.translation.translate('clipboard.error'),
+      });
+    }
+  }
+}
```

Comment:
praise: the whole thing works quite well - copying, translations and showing a toast.

suggestion: looking at the implementation as a whole, it seems like it's taken on a few responsibilities, which might be better off to be somewhere else. There's the copy functionality which is the core of the directive. Then there's a translation layer, and a notification/toast functionality.

I'm thinking of a possible split. What if we let the translation layer, and the notification/toast be handled by the host app? DH in our case. So from `WattCopyToClipboardDirective` we expose an `Output` that is called with the result of `this.clipboard.copy`. Then the host app will have to decide how to notify the user about the copy being successful/unsuccessful.

This leads to another issue that the developers should handle the notification/toast part. In this case we can make it really easy for them by creating another directive specific to DH that is applied to the same element the `wattCopyToClipboard` is applied (let's theoretically call it `dhNotifyOnCopy`). This way the DH directive can inject the `WattCopyToClipboardDirective` instance and subscribe to its `Output`.

So the end result for the developer will be:

```html
<span wattCopyToClipboard dhNotifyOnCopy`>Copy text</div>
```

If we feel adventurous enough, we can also use a `wattCopyToClipboard` selector in the DH directive. This way the `dhNotifyOnCopy` will not be needed at all on the element.

Advantages of this solution:
- no custom translation layer in the copy directive
- no extra translation files to maintain
- existing translation layer can be used
- each app decides how to notify the user on copy
- it's still easy to show a toast on copy by creating the extra directive

What do you think?

Reply:
I'm not a huge fan of having multiple directives for this case. If the problem is having multiple translation files or the fact that transloco is used inside of Watt my suggestion would be to configure the clipboard by just defining a token just like we set the locale of the date picker.

`
@NgModule({
  providers: [
    {provide: WATT_CLIPBOARD_SOMETHING_SOMETHING, useValue: ...},
  ],
})
export class MyApp {}
`